src/psyclone/psyir/transformations/reference2arrayrange_trans.py

```diff
@@ -146,6 +146,16 @@ def validate(self, node, options=None):
             raise TransformationError(
                 f"The supplied node should be a Reference to a symbol "
                 f"that is an array, but '{node.symbol.name}' is not.")
+        if isinstance(node.parent, BinaryOperation):
+            if node.ancestor(Loop):
+                # This is an array reference within an LBOUND or
+                # UBOUND binaryoperator.
+                if node.parent.operator in [
+                        BinaryOperation.Operator.LBOUND,
+                        BinaryOperation.Operator.UBOUND]:
+                    raise TransformationError(
```

Comment:
Please update the :raises: section of the doc string. Also, `SIZE` is a query function too.

Reply:
Possibly, we should also disallow the case where the Reference is an argument to a Call?

Reply:
Updated raises and added a test for SIZE.

Reply:
I'm not sure about Calls (and functions) as I think `call x(a)` and `call x(a(:)` are the same, as are `dotproduct(a,b)` and `dotproduct(a(:),b(:))` so it does not matter, and perhaps we want to be able to do this in some circumstances?.

Further, this PR is meant to be about adding the existing transformation to PSyAD, not modifying the transformation, so it could be a new issue if it is decided that is is an issue. I've had to make some changes to the existing transformation as integrating it into psyad showed that the transformation should not modify arrays within LBOUND, UBOUND and potentially (as you point out) SIZE, but I'm not sure we should be doing more here.

I'll leave it to the reviewer to decide the best way forward,

Reply:
I added a test, but left it at that.

Reply:
That sounds fine - I think I was getting confused.